### PR TITLE
gtk/textctrl: remove frame via CSS on recent gtk

### PR DESCRIPTION
**Pull request by adrientetar**
Background:
https://groups.google.com/forum/#!topic/wx-users/Wg8M-jC_8lk

Here's how it looks:
image

I also had to set padding and border radius to 0 otherwise the text got clipped.

r? @vadz

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,6 +11,13 @@
 /*tags
 /autom4te.cache
 
+# generated files
```
Comment (vadz) [inline .gitignore]: This shouldn't be part of this PR.

File: samples/widgets/textctrl.cpp [hunk 2]
```diff
@@ -64,6 +64,7 @@ enum
 
     TextPage_Password,
     TextPage_NoVertScrollbar,
+    TextPage_NoBorder,
```
Comment (vadz) [inline samples/widgets/textctrl.cpp]: I don't think we need to modify the sample, it already has "Set border" submenu which should allow testing with setting any border style for any of the controls, including the text control.

File: src/gtk/textctrl.cpp [hunk 3]
```diff
@@ -728,7 +728,26 @@ bool wxTextCtrl::Create( wxWindow *parent,
...
+            GtkCssProvider* provider = gtk_css_provider_new();
+            gtk_css_provider_load_from_data(
+                provider,
+                gtk_check_version(3,20,0) ? "entry" : "GtkEntry"
+                " { border: none; border-radius: 0; padding: 0 }",
+                -1, NULL);
+            gtk_style_context_add_provider(
+                gtk_widget_get_style_context(m_widget),
+                GTK_STYLE_PROVIDER(provider),
+                GTK_STYLE_PROVIDER_PRIORITY_USER);
```
Comment (vadz) [inline src/gtk/textctrl.cpp]: I don't think this is the right priority to use, `GTK_STYLE_PROVIDER_PRIORITY_APPLICATION` is probably right.

File: src/gtk/textctrl.cpp [hunk 4]
```diff
@@ -728,7 +728,26 @@ bool wxTextCtrl::Create( wxWindow *parent,
         gtk_entry_get_text((GtkEntry*)m_text);
 
         if (style & wxNO_BORDER)
-            g_object_set (m_text, "has-frame", FALSE, NULL);
+        {
+#ifdef __WXGTK3__
+            // this is sort of a workaround for when the builtin theme
+            // won't remove the frame by itself -- see
+            // https://bugzilla.gnome.org/show_bug.cgi?id=789732#c1
+            GtkCssProvider* provider = gtk_css_provider_new();
```
Comment (vadz) [inline src/gtk/textctrl.cpp]: Could we reuse the existing (but, curiously, currently unused) `wxWindowGTK::ApplyCssStyle()`?

File: src/gtk/textctrl.cpp [hunk 5]
```diff
@@ -728,7 +728,26 @@ bool wxTextCtrl::Create( wxWindow *parent,
...
+            gtk_css_provider_load_from_data(
+                provider,
+                gtk_check_version(3,20,0) ? "entry" : "GtkEntry"
```
Comment (vadz) [inline src/gtk/textctrl.cpp]: Sorry, this can't be right. A run-time `gtk_check_version()` check doesn't work well with compile-time string concatenation, this will use just "entry" (and nothing else) for pre-3.20 versions.

**Review by vadz — CHANGES_REQUESTED** [review]
Thanks for working on this, but a few things would need to be changed for this to be merged.

File: .gitignore (hunk 1 above)
Comment (adrientetar) [reply .gitignore]: You want a separate PR for it?

Comment (vadz) [reply .gitignore]: I'd just avoid changing this at all, you can always put this in your `.git/info/exclude`.

And, FWIW, I strongly advise to always build outside of the source directory, then you won't have these files in the repository at all.

**Comment by adrientetar** [comment]
Alright, this should be good to go.

File: src/gtk/textctrl.cpp [hunk 6]
```diff
@@ -728,7 +728,21 @@ bool wxTextCtrl::Create( wxWindow *parent,
...
+#ifdef __WXGTK3__
+            // this is sort of a workaround for when the builtin theme
+            // won't remove the frame by itself -- see
+            // https://bugzilla.gnome.org/show_bug.cgi?id=789732#c1
+            GtkCssProvider* provider = gtk_css_provider_new();
+            std::string rule = !gtk_check_version(3,20,0) ? "entry" : "GtkEntry";
+            rule.append(" { border: none; border-radius: 0; padding: 0 }");
+
+            wxWindowGTK::ApplyCssStyle(provider, rule.c_str());
+            g_object_unref(provider);
```
Comment (vadz) [inline src/gtk/textctrl.cpp]: Could we please use `wxGtkObject` from `include/wx/gtk/private/object.h` instead of manual `g_object_unref()`?

This may seem like needless complication for such a simple piece of code, but simple code has tendency to become less simple with time and it could well happen that someone introduces a `return` somewhere above resulting in a memory leak with the code as written.

**Comment by vadz** [comment]
Thanks for the update, I agree that this should be committed (with just one small change, as per above), but I thought about something else: shouldn't we actually do this for all controls, not just the text one? E.g. it looks like there is a similar problem [with wxButton](https://trac.wxwidgets.org/ticket/17683).

So maybe what we really need is a virtual `GTKGetCssSelector()` method (maybe "selector" is not the right term to use here, but I don't know what is it called) that would return "entry" or "GtkEntry", but also "button or "GtkButton" etc, and handle this generically at `wxControl` level?

**Comment by adrientetar** [comment]
> E.g. it looks like there is a similar problem with wxButton.

I just tried in the widgets sample and border=none works for wxButton, the bug mentions wxBitmapButton which I don't see in the sample, but if it's a subclass of wxButton I think it should work.

image

Then again maybe this class of bugs depends on the system theme, it's hard to tell.

By the way in the sample I noticed the spin controls have sizing issues.

image

~~~
(widgets:13289): Gtk-CRITICAL **: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkSpinButton
~~~

**Review by vadz — APPROVED** [review]

File: src/gtk/textctrl.cpp [hunk 7]
```diff
@@ -728,7 +729,20 @@ bool wxTextCtrl::Create( wxWindow *parent,
...
 
         if (style & wxNO_BORDER)
-            g_object_set (m_text, "has-frame", FALSE, NULL);
+        {
+#ifdef __WXGTK3__
+            // this is sort of a workaround for when the builtin theme
+            // won't remove the frame by itself -- see
+            // https://bugzilla.gnome.org/show_bug.cgi?id=789732#c1
+            wxGtkObject<GtkCssProvider> provider(gtk_css_provider_new());
+            std::string rule = !gtk_check_version(3,20,0) ? "entry" : "GtkEntry";
```
Comment (paulcor) [inline src/gtk/textctrl.cpp]: There's no need to name the selector, just use "*".

File: src/gtk/textctrl.cpp [hunk 8]
```diff
@@ -728,7 +729,20 @@ bool wxTextCtrl::Create( wxWindow *parent,
...
+            rule.append(" { border: none; border-radius: 0; padding: 0 }");
+
+            wxWindowGTK::ApplyCssStyle(provider, rule.c_str());
```
Comment (paulcor) [inline src/gtk/textctrl.cpp]: Don't need the `wxWindowGTK::` prefix.
